### PR TITLE
extractMatchedPath directive

### PR DESCRIPTION
**Pull request by rbudzko**
- Directive itself based on @2beaucoup suggestion
- Used in directory path managing
- Tests + Paradox

Refs #496.

### CODE REVIEW

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/BasicDirectives.scala [hunk 1]
```diff
@@ -375,4 +384,19 @@ object BasicDirectives extends BasicDirectives {
   private val _extractRequestContext: Directive1[RequestContext] = extract(scalaIdentityFunction)
   private val _extractRequestEntity: Directive1[RequestEntity] = extract(_.request.entity)
   private val _extractDataBytes: Directive1[Source[ByteString, Any]] = extract(_.request.entity.dataBytes)
+
+  private def extractMatched(ctx: RequestContext) = {
+    val unmatchedPath = ctx.unmatchedPath
+    val fullPath = ctx.request.uri.path.toString
+    val matchedLength = fullPath.lastIndexOf(unmatchedPath.toString)
```
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/BasicDirectives.scala]:
```scala
require(
  fullPath.endsWith(unmatchedPath),
  ...
)
Path(fullPath.substring(0, fullPath.length - unmatchedPath.length))
```
?

File: docs/src/main/paradox/java/http/routing-dsl/directives/basic-directives/extractUnmatchedPath.md [hunk 2]
```diff
@@ -11,6 +11,8 @@ that handle complete suffixes of paths (like the `getFromDirectory` directives a
 
 Use `mapUnmatchedPath` to change the value of the unmatched path.
 
+See also @ref[extractMatchedPath](extractMatchedPath.md#extractmatchedpath-java) to see similar directive for matched path.
```
Comment (2beaucoup) [inline docs/src/main/paradox/java/http/routing-dsl/directives/basic-directives/extractUnmatchedPath.md]: 👍

File: docs/src/main/paradox/java/http/routing-dsl/directives/basic-directives/extractMatchedPath.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<a id="extractmatchedpath-java"></a>
+# extractMatchedPath
+
+## Description
+
+Extracts the request's matched path.
+
+This directive is used to acquire part of path matched during routing matching
+process. Similar but opposite to `RequestContext` unmatched path property.
```
Comment (2beaucoup) [inline docs/src/main/paradox/java/http/routing-dsl/directives/basic-directives/extractMatchedPath.md]: Wording could be slightly improved. Just steal + adapt from `extractUnmatchedPath`?

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/BasicDirectives.scala (hunk 1 above)
Comment (rbudzko) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/BasicDirectives.scala]: Yes, definitely.

**Comment by rbudzko** [comment]
@2beaucoup Done.

**Comment by jypma** [comment]
Good stuff! LGTM

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileAndResourceDirectives.scala [hunk 4]
```diff
@@ -146,26 +146,25 @@ trait FileAndResourceDirectives {
...
-        val path = ctx.unmatchedPath
-        val fullPath = ctx.request.uri.path.toString
-        val matchedLength = fullPath.lastIndexOf(path.toString)
-        require(matchedLength >= 0, s"Unmatched path '$path' wasn't a suffix of full path '$fullPath'. " +
-          "This usually means that ctx.unmatchedPath was manipulated inconsistently " +
-          "with ctx.request.uri.path")
-        val pathPrefix = fullPath.substring(0, matchedLength)
-        val pathString = withTrailingSlash(safeJoinPaths("/", path, ctx.log, '/'))
-        val dirs = directories flatMap { dir ⇒
-          safeDirectoryChildPath(withTrailingSlash(dir), path, ctx.log) match {
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileAndResourceDirectives.scala]: I'm getting chills that we are changing this method again... It LGTM but can someone run the tests on Windows to make sure we tried at least not to introduce another hole here?

**Comment by jrudolph** [comment]
LGTM. Can someone run the tests on Windows?

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileAndResourceDirectives.scala (hunk 4 above)
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileAndResourceDirectives.scala]: Agreed. I think @rbudzko is on Windows but I'll double check. The behavior shouldn't have change here as it's just a refactoring (famous last words...)

**Comment by rbudzko** [comment]
I've run tests (I mean `sbt test`) on Windows. When I think about it right now I should have created simple web app as well to test it by hand. I can do it on Monday, because now is a weekend and weekend means linux 😀 .

**Comment by 2beaucoup** [comment]
Tests pass for me on Windows as well.

**Comment by jrudolph** [comment]
Great, thanks @2beaucoup and @rbudzko for running the tests (and yay to Linux on weekends! Still better than no Linux at all ;) ).

**Comment by ktoso** [comment]
Very nice one, thanks @rbudzko 👍

**Pull request merged**